### ISSUE TITLE
method Generate with Async

### ISSUE THREAD

**Issue by drma-dev** [bchavez/Bogus#345]
When using the method to generate a list there is usually a certain delay. Is it possible to implement the async / await version?

**Comment by bchavez** [comment]
Hello,

I'm not sure I see much benefit of using `async`/`await` in **Bogus**. Almost all the work in **Bogus** is CPU bound. There isn't much I/O bound work in **Bogus**.

If you have ideas on *specific parts* of **Bogus** to `await`; let me know.

Otherwise, if you are seeing a delay; I would say the list is too large for a single thread to process quickly on a single CPU core. In such cases, consider breaking up and partitioning the work if you have additional CPU cores to provide.

For example, consider the various ways to generate 5 million fake unique emails, using Threads, Tasks, and `Parallel.For`:

```csharp
public string[] GetEmails(int start, int count)
{
   var f = new Faker()
   {
      Random = new Randomizer(start)
   };

   var emails = Enumerable.Range(0, count)
      .Select(i =>
      {
         var uniqueId = (start + i).ToString();
         return f.Internet.Email(uniqueSuffix: uniqueId);
      }).ToArray();

   return emails;
}
```

-----

### Using a Single Thread

Using one thread on a computer with 12 CPU cores:

```csharp
void Main()
{
   var sw = Stopwatch.StartNew();
   
   var allEmails = GetEmails(0, 5_000_000);
   $"Email Count: {allEmails.Length:n0}".Dump();
   allEmails.Take(5).Dump()
   
   sw.Stop();
   sw.Elapsed.Dump();
}
```
```
Email Count: 5,000,000
----------------------
Reanna.Runte0@hotmail.com
Susie.Kihn291@yahoo.com
Lora.Kovacek862@yahoo.com
Marianna_Haley993@hotmail.com
Maximus_Leffler544@gmail.com
...
00:00:16.0371913
```

-----

### Using Five Tasks

Using 5 tasks, each task with 1 million generations, on a computer with 12 CPU cores:

```csharp
async Task Main()
{
   var sw = Stopwatch.StartNew();
   
   var t1 = Task.Run(() => GetEmails(0, 1_000_000));
   var t2 = Task.Run(() => GetEmails(1_000_001, 1_000_000));
   var t3 = Task.Run(() => GetEmails(2_000_001, 1_000_000));
   var t4 = Task.Run(() => GetEmails(3_000_001, 1_000_000));
   var t5 = Task.Run(() => GetEmails(4_000_001, 1_000_000));

   var tasks = new List<Task<string[]>>
   {
      t1, t2, t3, t4, t5
   };
   await Task.WhenAll(tasks);
   
   var allEmails = new List<string>();
   foreach( var t in tasks){
      var batch = t.GetAwaiter().GetResult();
      allEmails.AddRange( batch );
   }

   $"Email Count: {allEmails.Count}".Dump();
   allEmails.Take(5).Dump();

   sw.Stop();
   sw.Elapsed.Dump();
}
```
```
Email Count: 5,000,000
----------------------
Reanna.Runte0@hotmail.com
Susie.Kihn291@yahoo.com
Lora.Kovacek862@yahoo.com
Marianna_Haley993@hotmail.com
Maximus_Leffler544@gmail.com
...
00:00:06.1459384
```

-----

### Using `Parallel.For`

Using `Parallel.For` on a computer with 12 CPU cores:

```csharp
void Main()
{   
   var seed = 0;
   var bag = new ConcurrentBag<List<string>>();

   var sw = Stopwatch.StartNew();

   var result = Parallel.For(0, 5_000_000, () => CreateLoopState(ref seed),
         (i, state, local)=>{
            
            var email = local.faker.Internet.Email(uniqueSuffix: i.ToString());
            local.batch.Add(email);
            
            return local;
         }
      ,
         (local) => {
            bag.Add(local.batch);
         }
      );
   
   var allEmails = bag.ToArray().SelectMany(batch => batch).ToArray();
   $"Email Count: {allEmails.Length:n0}".Dump();
   allEmails.Take(5).Dump();

   sw.Stop();
   sw.Elapsed.Dump();
}

public static (Faker faker, List<string> batch) CreateLoopState(ref int seed)
{
   var localSeed = Interlocked.Increment(ref seed);
   var f = new Faker()
   {
      Random = new Randomizer(localSeed)
   };
   return (f, new List<string>());
}
```
```
Email Count: 5,000,000
----------------------
Bernita_Konopelski430@gmail.com
Tyree.Bradtke1@yahoo.com
Emmet_Zboncak282@gmail.com
Mckenna.Prohaska1619@yahoo.com
Pierce.Dach8220@yahoo.com
...
00:00:08.3452107
```

-----

In summary, the amount and *degree* of parallelization is up to you. The ultimate throughput you get from **Bogus** depends on your hardware configuration. **Bogus** can't and *won't try* to break up or partition the work for you. You have to make a decision on what is best for your scenario.

The only guidance I can provide is:
* If you want to `async`-ify **Bogus** consider using `Task.Run(() => faker.Generate())`
* Each thread should have a dedicated instance of `Faker`/`Faker<T>`/`DataSet` and a locally set `Seed` or `Randomizer`.

Hope that helps.

Thanks,
Brian

**Comment by drma-dev** [comment]
@bchavez 

I am testing only with 12 records, but it is related to the functions that generate image url. I believe that having a native async, would facilitate a lot.

first run: 00:00:03.4131600
second run: 00:00:01.1827150

```
return new Faker<ProfilePhoto>("pt_BR")
                .Rules((s, p) =>
                {
                    p.UpdateMainPhoto(s.Internet.Avatar());
                    p.UpdatePhotoGallery(new[] { s.Image.PicsumUrl() });
                });
```

**Comment by bchavez** [comment]
@drma-dev Are you making database calls in `.UpdateMainPhoto()` and `.UpdatePhotoGallery()` with **Bogus** within the `.Rules()` method?

**Comment by drma-dev** [comment]
No. I am populating properties.

```
public class ProfilePhoto
    {
        public string Main { get; private set; }
        public string Validation { get; private set; }
        public string[] Gallery { get; private set; } = Array.Empty<string>();

        public void UpdateMainPhoto(string Main)
        {
            this.Main = Main;
        }

        public void UpdatePhotoGallery(string[] Gallery)
        {
            this.Gallery = Gallery;
        }

        public string GetMainPhoto()
        {
            if (string.IsNullOrEmpty(Main))
                return "/img/nouser.jpg";
            else
                return Main;
        }
    }
```

**Comment by bchavez** [comment]
@drma-dev you must be doing some extra processing somewhere in your application. I created **10,000** `ProfilePhoto` and took ~18-20 milliseconds:

```csharp
void Main()
{
   var sw = Stopwatch.StartNew();
   var f = new Faker<ProfilePhoto>("pt_BR")
                .Rules((s, p) =>
                {
                   p.UpdateMainPhoto(s.Internet.Avatar());
                   p.UpdatePhotoGallery(new[] { s.Image.PicsumUrl() });
                });
                
   var profilePhotos = f.Generate(10_000);
   
   sw.Stop();
   sw.ElapsedMilliseconds.Dump();
   sw.Elapsed.Dump();
   
   profilePhotos.Dump();
}

public class ProfilePhoto
{
   public string Main { get; private set; }
   public string Validation { get; private set; }
   public string[] Gallery { get; private set; } = Array.Empty<string>();

   public void UpdateMainPhoto(string Main)
   {
      this.Main = Main;
   }

   public void UpdatePhotoGallery(string[] Gallery)
   {
      this.Gallery = Gallery;
   }

   public string GetMainPhoto()
   {
      if (string.IsNullOrEmpty(Main))
         return "/img/nouser.jpg";
      else
         return Main;
   }
}
```

![image](https://user-images.githubusercontent.com/478118/103444975-72e35700-4c23-11eb-8af9-50670dd4a35f.png)

You might want to check your application if it is doing extra work somewhere.

**Comment by drma-dev** [comment]
I have a suspicion, but I'm not sure. What is the best way to popular class attributes?

```
public static Faker<ProfileSearch> GetProfileSearch(string Id = null)
        {
            return new Faker<ProfileSearch>("pt_BR")
                .Rules((s, p) =>
                {
                    p.SetIds(Id ?? s.Random.Guid().ToString());
                    p.NickName = s.Name.FirstName();
                    p.BirthDate = s.Date.Past(18).Date;
                    p.UpdatePhoto(GetProfilePhoto());
                    //p.UpdatePhoto(new ProfilePhoto());
                    p.ActivityStatus = s.PickRandom<ActivityStatus>();
                    p.Distance = s.Random.Number(500, 10000);
                });
        }
```
```
public static Faker<ProfilePhoto> GetProfilePhoto()
        {
            return new Faker<ProfilePhoto>("pt_BR")
                .Rules((s, p) =>
                {
                    p.UpdateMainPhoto(s.Internet.Avatar());
                    p.UpdatePhotoGallery(new[] { s.Image.PicsumUrl() });
                });
        }
```

I am creating two Fake objects in sequence.

**Comment by bchavez** [comment]
@drma-dev I wish I could help, but I think you are the best one to find the answer. I don't know what your program does. I don't have access to your code. I can only see small clips of how **Bogus** is being used.

It's like being asked to solve a crime when only provided three photos. You need to do *testing* to find the problem. Additionally, you have an asymmetrical information ***advantage*** because you can also view **Bogus'** source code to help you find the problem.

If you find a problem with **Bogus** then you have to present a test case that shows the problem with **Bogus** only, not your code + **Bogus**.
